sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetFileFormat.scala

```diff
@@ -105,26 +105,60 @@ object GpuParquetFileFormat {
     }
   }
 
-  def parquetWriterOptionsFromSchema[T <: NestedBuilder[_, _], V <: ParquetColumnWriterOptions]
-  (builder: ParquetColumnWriterOptions.NestedBuilder[T, V],
-   schema: StructType, writeInt96: Boolean): T = {
+  def getParquetColumnWriterOption(
+      name: String,
+      dataType: DataType,
+      writeInt96: Boolean,
+      nullable: Boolean): ParquetColumnWriterOptions = {
+    dataType match {
+      case dt: DecimalType =>
+        // key can never be null
+        new ParquetColumnWriterOptions(name, writeInt96, dt.precision, nullable)
+      case TimestampType =>
+        new ParquetColumnWriterOptions(name, writeInt96, 0, nullable)
+      case s: StructType =>
+        parquetWriterOptionsFromSchema(structBuilder(name, nullable), s, writeInt96).build()
+      case a: ArrayType =>
+        parquetWriterOptionsFromSchema(listBuilder(name),
+          StructType(Array(StructField(name, a.elementType, nullable))), writeInt96).build()
+      case m: MapType =>
+        mapColumn(name,
+          getParquetColumnWriterOption("key", m.keyType, writeInt96, false),
+          getParquetColumnWriterOption("value", m.valueType, writeInt96, nullable))
+      case _ =>
+        new ParquetColumnWriterOptions(name, nullable)
+    }
+  }
+
+  def parquetWriterOptionsFromSchema[T <: NestedBuilder[_, _], V <: ParquetColumnWriterOptions](
+      builder: ParquetColumnWriterOptions.NestedBuilder[T, V],
+      schema: StructType,
+      writeInt96: Boolean,
+      nullable: Boolean = true): T = {
```

Comment:
I really dislike having a parameter for nullable separate from the schema.  The schema should have nullable in it. If it is a `nullableOverride`, then name it like that. We need comments explaining how this is supposed to be used and what it does. Because from what I can see there are bugs in this code. `Map<Array<String>, String>` with the current code says not only that the key to the Map needs to not be nullable, but also that the Strings within the Array cannot be nullable, which I don't think is true.  But I could be wrong, because the recursion appears to call into `getParquetColumnWriterOption` for map keys and values only and honestly trying to trace this through everywhere with the separate functions that appear to be 99% the same is really confusing to me.